### DISCUSSION TITLE
Modify MemberNotNullAttribute to be applied to constructors (add AttributeTargets.Constructor)

### DISCUSSION THREAD

**Discussion by LinoBarreca** [dotnet/csharplang#8831] [Language Ideas]
I read the issue about the MemberNotNullAttribute (dotnet/runtime#33567 & dotnet/roslyn#41438)
I don't understand why the developer is forced to write an Init() method (to which he should apply the MemberNotNullAttribute) instead of applying the MemberNotNullAttribute to the constructor itself.
If my constructor ensured the member of the class is never null why can't I apply the MemberNotNullAttribute to the contructor itself specifying the member name in the attribute?
Having a constructor that must call another method (with the attribute) looks like a nonsense to me.
I understand the CS8618 if the constructor exits and the member may not have been initialized but why can't be the constructor itself ensuring the member is not null?

Please add the AttributeTargets.Constructor to MemberNotNullAttribute/MemberNotNullWhenAttribute

[a comment by an automated account is not shown]

**Comment by huoyaoyuan** [comment]
Shouldn't `MemberNotNull` on constructor just be a non-nullable member?
I remember that the nullablity verification of constructor is going go be changed to more close to MemberNotNull`.

**Comment by gfoidl** [comment]
Another usecase for this would be something like
```c#
public class MyWorker 
{
    private static Task<Dictionary<string, object>>? s_mappingTask;
    
    public MyWorker(IService service)
    {
        s_mappingTask ??= service.GetMapAsync();
    }
    
    public async ValueTask<object> UseObjectFromCachedMapping(string key)
    {
        Dictionary<string, object> mapping = await s_mappingTask;    // <-- warning CS8602: Dereference of a possibly null reference.
        return mapping[key];    // no check here for demo purposes
    }
}

public interface IService
{
    Task<Dictionary<string, object>> GetMapAsync();
}
```
Applying `[MemberNotNull(nameof(s_mappingTask))]` to the ctor would fix the warning.
(I know there are other ways, too).

**Comment by webczat** [comment]
use case: structs. struct can always be defaulted, so the reference type field would be null, so you'd mark these fields nullable. But it would be nice to be able to mark declared constructors as ensuring the fields are not null when these constructors were used.

**Comment by 333fred** [comment]
The problem with that use case is that it falls apart immediately when you pass the struct anywhere, which is no longer sure that the struct was validly constructed. Non-defaultable struct types are a large feature on their own.

**Comment by webczat** [comment]
you may have some other way to check whether the struct has a valid value. Like in general if default is not acceptable somewhere you should check for this anyway in everything that you pass the struct to, right? and you need that even if you have nrt off. so it doesn't change anything, but allows me to say that if I use this constructor, these fields are not null. if I pass the struct to some method, the method would have to check for that case anyway if it's invalid, even if nrts didn't exist. And if that case is valid, I don't need any annotation.

**Comment by webczat** [comment]
note I have MemberNotNullWhen, and I can use that, or check for specific nullable field values, in case I got the struct from external source.
Oh the problem happens if I really mean default as invalid in some context and some method is guaranteed to return a valid non default struct, then I need to check unnecessarily or silence nullability warnings.

**Comment by jnm2** [comment]
@333fred I think the ask is to be able to avoid an IsValid check immediately after calling a struct constructor, even though you'd have to redo the IsValid check wherever you pass the struct. (Where IsValid is a bool property with MemberNotNullWhen connected to a bunch of nullable reference properties.)

**Comment by webczat** [comment]
for my use case that's the thing. But even without such a bool property you could just check stuff for null if youdon't know whether what you get is default or not, but when you just constructed the thing then it's known that the fields are not null at the moment.
Definitely doesn't cover all cases, but helps a bit.
